### DISCUSSION TITLE
Atomic Guards, Unique reservations and SingleNode optimistic concurrency with long lived sessions

### DISCUSSION THREAD

**Discussion by mcrio** [ravendb/ravendb#16519] [General]
In [this article](https://ravendb.net/articles/ravendb-5-2-simplifying-atomic-cluster-wide-transactions) it was discussed to use reservation documents for unique value reservations and cluster-wide transactions with atomic guards.

```
using var session = store.OpenAsyncSession(new SessionOptions
{
    TransactionMode = TransactionMode.ClusterWide
});

var user = new User{Name = "Ayende"};
await session.StoreAsync(user);
await session.StoreAsync(new {ReservedFor = user.Id}, "usernames/" + user.Name);

await session.SaveChangesAsync();
```

Also it was suggested to use cluster wide transactions when we need to update the unique values (username, email) and to use single-node transactions for all other updates.

> A common scenario for this code above is that you’ll include the usernames/ayende reservation document whenever you modify the user’s name, and that will always be using a cluster wide transactions. However, when modifying the user and not the user name, you can skip the cost of a cluster wide transaction.

How does this apply to desktop applications that keep a session per form, like suggested by Oren in the following article: [https://learn.microsoft.com...](https://disq.us/url?url=https%3A%2F%2Flearn.microsoft.com%2Fen-us%2Farchive%2Fmsdn-magazine%2F2009%2Fdecember%2Fdata-access-building-a-desktop-to-do-application-with-nhibernate%3Aym4M5fwEZgJPB9CeG9KCXusazls&cuid=5987129)

Asking as in the Desktop scenario, a form opened to update the unique username and maybe some other user details, that form will overwrite any other update that happened in meantime and which did not include a change of a unique value (cluster-wide). 

Seems like in the Desktop app case all the transactions would need to be cluster-wide, as the atomic guard change vector does not get updated when a single node transaction settles?

**Comment by garayx** [comment]
yes, if you would like to update the unique values the session have to be cluster wide, I guess if you are using cluster wide tx, its better to use session-per-request style,

**Reply by mcrio** [reply]
Thanks. As described in Oren's article linked above, session per request is good when doing web applications, but for desktop applications you'd lose many benefits of the session like change tracking etc. 

I was just interesting in what the recommended way of doing things would be with cluster-wide atomic guards and "long-lived" sessions.

**Reply by ayende** [reply]
A session per form in a desktop application can still use cluster wide tx.
Note that the transaction applies *during* the `SaveChanges` call, not for the lifetime of the form, so the usual caveats about keeping it open for a long time do not apply
